File: scripts/scrape-agent-urls.js
const { PrismaClient } = require('@prisma/client')
const axios = require('axios')
const cheerio = require('cheerio')

const prisma = new PrismaClient()

const BASE_URL = "https://www.transfermarkt.es"

const headers = {
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
}

async function getAgentByPlayer(name) {
  try {
    console.log(`🔍 Buscando jugador: ${name}`)
    
    // 1. Buscar el jugador en Transfermarkt
    const searchUrl = `${BASE_URL}/schnellsuche/ergebnis/schnellsuche`
    const searchResponse = await axios.get(searchUrl, {
      params: { query: name },
      headers,
      timeout: 10000
    })
    
    const $ = cheerio.load(searchResponse.data)
    
    // 2. Tomar el primer resultado de jugador
    // Intentar diferentes selectores
    let playerLink = $('a.spielprofil_tooltip').first()
    if (playerLink.length === 0) {
      playerLink = $('a[href*="/profil/spieler/"]').first()
    }
    if (playerLink.length === 0) {
      playerLink = $('a[href*="/spieler/"]').first()
    }
    
    if (playerLink.length === 0) {
      console.log(`❌ No se encontró jugador: ${name}`)
      console.log('🔍 HTML de búsqueda:', $.html().substring(0, 1000))
      return { playerUrl: null, agentInfo: null }
    }
    
    const playerHref = playerLink.attr('href')
    const playerUrl = BASE_URL + playerHref
    console.log(`✅ Jugador encontrado: ${playerUrl}`)
    
    // 3. Entrar al perfil del jugador
    const playerResponse = await axios.get(playerUrl, {
      headers,
      timeout: 10000
    })
    
    const $player = cheerio.load(playerResponse.data)
    
    // 4. Buscar la sección "Agente" - métodos mejorados
    let agentInfo = null
    
    // Método 1: Buscar en la sección de datos principales (.data-header__label)
    const dataHeaderLabel = $player('.data-header__label').filter(function() {
      return $(this).text().trim() === 'Agente:'
    }).first()
    
    if (dataHeaderLabel.length > 0) {
      const agentLink = dataHeaderLabel.find('a').first()
      if (agentLink.length > 0) {
        const agentName = agentLink.text().trim()
        const agentUrl = BASE_URL + agentLink.attr('href')
        agentInfo = { name: agentName, url: agentUrl }
      }
    }
    
    // Método 2: Buscar en la tabla de información (.info-table__content)
    if (!agentInfo) {
      const infoTableContent = $player('.info-table__content').filter(function() {
        return $(this).text().trim() === 'Agente:'
      }).first()
      
      if (infoTableContent.length > 0) {
        const agentLink = infoTableContent.next('.info-table__content').find('a').first()
        if (agentLink.length > 0) {
          const agentName = agentLink.text().trim()
          const agentUrl = BASE_URL + agentLink.attr('href')
          agentInfo = { name: agentName, url: agentUrl }
        }
      }
    }
    
    // Método 3: Buscar por texto "Agent:" (inglés)
    if (!agentInfo) {
      const agentSpan = $player('span').filter(function() {
        return $(this).text().trim() === 'Agent:'
      }).first()
      
      if (agentSpan.length > 0) {
        const agentLink = agentSpan.next('a')
        if (agentLink.length > 0) {
          const agentName = agentLink.text().trim()
          const agentUrl = BASE_URL + agentLink.attr('href')
          agentInfo = { name: agentName, url: agentUrl }
        }
      }
    }
    
    // Método 4: Buscar enlaces específicos de agentes (no páginas generales)
    if (!agentInfo) {
      const agentLinks = $player('a[href*="/berater/"]').filter(function() {
        const href = $(this).attr('href')
        // Excluir páginas generales de estadísticas
        return href && !href.includes('beraterfirmenuebersicht') && !href.includes('statistik')
      })
      
      if (agentLinks.length > 0) {
        const agentLink = agentLinks.first()
        const agentName = agentLink.text().trim()
        const agentUrl = BASE_URL + agentLink.attr('href')
        agentInfo = { name: agentName, url: agentUrl }
      }
    }
    
    if (agentInfo) {
      console.log(`✅ Agente encontrado: ${agentInfo.name} - ${agentInfo.url}`)
      return { playerUrl, agentInfo }
    } else {
      console.log(`ℹ️ Jugador sin agente registrado: ${name}`)
      return { playerUrl, agentInfo: null }
    }
    
    return { playerUrl, agentInfo: null }
    
  } catch (error) {
    console.error(`❌ Error para ${name}:`, error.message)
    return { playerUrl: null, agentInfo: null }
  }
}

async function scrapePlayersFromDatabase() {
  try {
    console.log('🚀 Iniciando scraping de URLs de agentes...')
    
    // Obtener todos los jugadores que no tienen URL de agente
    const players = await prisma.jugador.findMany({
      where: {
        OR: [
          { url_trfm_advisor: null },
          { url_trfm_advisor: '' }
        ]
      },
      select: {
        id_player: true,
        player_name: true,
        complete_player_name: true
      },
      take: 50 // Limitar a 50 jugadores por ejecución para evitar timeouts
    })
    
    console.log(`📊 Encontrados ${players.length} jugadores sin URL de agente`)
    
    for (let i = 0; i < players.length; i++) {
      const player = players[i]
      const fullName = player.complete_player_name || player.player_name
      
      console.log(`\n[${i + 1}/${players.length}] Procesando: ${fullName}`)
      
      // Obtener información del agente
      const { playerUrl, agentInfo } = await getAgentByPlayer(fullName)
      
      // Actualizar la base de datos
      const updateData = {}
      if (playerUrl) {
        updateData.url_trfm_advisor = playerUrl
      }
      
      if (agentInfo) {
        updateData.url_trfm_advisor = agentInfo.url
      }
      
      if (Object.keys(updateData).length > 0) {
        await prisma.jugador.update({
          where: { id_player: player.id_player },
          data: updateData
        })
        console.log(`✅ Actualizado en BD:`, updateData)
      } else {
        console.log(`❌ No se pudo obtener información para: ${fullName}`)
      }
      
      // Pausa aleatoria para evitar ser bloqueado
      const delay = Math.random() * 2000 + 1000 // Entre 1-3 segundos
      await new Promise(resolve => setTimeout(resolve, delay))
    }
    
    console.log('\n🎉 Scraping completado!')
    
  } catch (error) {
    console.error('❌ Error general:', error)
  } finally {
    await prisma.$disconnect()
  }
}

// Función para probar con un jugador específico
async function testPlayer(playerName) {
  console.log(`🧪 Probando con jugador: ${playerName}`)
  const result = await getAgentByPlayer(playerName)
  
  console.log('\n📋 Resultado:')
  console.log(`Perfil del jugador: ${result.playerUrl}`)
  if (result.agentInfo) {
    console.log(`Agente: ${result.agentInfo.name}`)
    console.log(`Enlace agente: ${result.agentInfo.url}`)
  } else {
    console.log('No tiene agente registrado.')
  }
}

// Ejecutar según los argumentos
if (process.argv.length > 2) {
  const playerName = process.argv.slice(2).join(' ')
  testPlayer(playerName)
} else {
  scrapePlayersFromDatabase()
}

File: scripts/scrape_agent_urls.py
import requests
from bs4 import BeautifulSoup
import time
import random
import os
import sys
from typing import Optional, Tuple
from dotenv import load_dotenv

BASE_URL = "https://www.transfermarkt.es"

headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                  "AppleWebKit/537.36 (KHTML, like Gecko) "
                  "Chrome/126.0.0.0 Safari/537.36"
}

def get_agent_by_player(name: str) -> Tuple[Optional[str], Optional[Tuple[str, str]]]:
    """
    Busca un jugador en Transfermarkt y obtiene su URL de perfil y agente.
    
    Args:
        name: Nombre del jugador a buscar
        
    Returns:
        Tuple con (player_url, agent_info) donde agent_info es (agent_name, agent_url) o None
    """
    try:
        # 1. Buscar el jugador en Transfermarkt
        search_url = f"{BASE_URL}/schnellsuche/ergebnis/schnellsuche"
        params = {"query": name}
        
        print(f"🔍 Buscando jugador: {name}")
        r = requests.get(search_url, params=params, headers=headers, timeout=10)
        r.raise_for_status()
        
        soup = BeautifulSoup(r.text, "html.parser")

        # 2. Tomar el primer resultado de jugador
        player_link = soup.select_one("a.spielprofil_tooltip")
        if not player_link:
            print(f"❌ No se encontró jugador: {name}")
            return None, None
        
        player_href = player_link["href"]
        player_url = BASE_URL + player_href
        print(f"✅ Jugador encontrado: {player_url}")

        # 3. Entrar al perfil del jugador
        r2 = requests.get(player_url, headers=headers, timeout=10)
        r2.raise_for_status()
        soup2 = BeautifulSoup(r2.text, "html.parser")

        # 4. Buscar la sección "Agente"
        agent_span = soup2.find("span", string="Agente:")
        if not agent_span:
            print(f"ℹ️ Jugador sin agente registrado: {name}")
            return player_url, None

        agent_link = agent_span.find_next("a")
        if agent_link:
            agent_name = agent_link.text.strip()
            agent_url = BASE_URL + agent_link["href"]
            print(f"✅ Agente encontrado: {agent_name} - {agent_url}")
            return player_url, (agent_name, agent_url)

        return player_url, None
        
    except requests.exceptions.RequestException as e:
        print(f"❌ Error de conexión para {name}: {e}")
        return None, None
    except Exception as e:
        print(f"❌ Error inesperado para {name}: {e}")
        return None, None

def scrape_players_from_database():
    """
    Obtiene todos los jugadores de la base de datos y actualiza sus URLs de agente.
    """
    # Cargar variables de entorno
    load_dotenv()
    
    # Agregar el directorio raíz al path para importar módulos
    sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    
    from prisma import PrismaClient
    
    # Conectar a la base de datos
    prisma = PrismaClient()
    
    try:
        # Obtener todos los jugadores que no tienen URL de agente
        players = prisma.jugador.find_many(
            where={
                "OR": [
                    {"url_trfm_advisor": None},
                    {"url_trfm_advisor": ""}
                ]
            },
            select={
                "id_player": True,
                "player_name": True,
                "complete_player_name": True
            }
        )
        
        print(f"📊 Encontrados {len(players)} jugadores sin URL de agente")
        
        for i, player in enumerate(players, 1):
            # Usar complete_player_name si está disponible, sino player_name
            full_name = player.get('complete_player_name') or player.get('player_name', '')
            print(f"\n[{i}/{len(players)}] Procesando: {full_name}")
            
            # Obtener información del agente
            player_url, agent_info = get_agent_by_player(full_name)
            
            # Actualizar la base de datos
            update_data = {}
            if player_url:
                update_data['url_trfm_advisor'] = player_url
            
            if agent_info:
                agent_name, agent_url = agent_info
                update_data['url_trfm_advisor'] = agent_url
            
            if update_data:
                prisma.jugador.update(
                    where={"id_player": player['id_player']},
                    data=update_data
                )
                print(f"✅ Actualizado en BD: {update_data}")
            else:
                print(f"❌ No se pudo obtener información para: {full_name}")
            
            # Pausa aleatoria para evitar ser bloqueado
            time.sleep(random.uniform(1, 3))
            
    except Exception as e:
        print(f"❌ Error general: {e}")
    finally:
        prisma.disconnect()

if __name__ == "__main__":
    # Ejemplo de uso individual
    if len(sys.argv) > 1:
        player_name = " ".join(sys.argv[1:])
        player_url, agent = get_agent_by_player(player_name)
        
        print(f"\n📋 Resultado para '{player_name}':")
        print(f"Perfil del jugador: {player_url}")
        if agent:
            print(f"Agente: {agent[0]}")
            print(f"Enlace agente: {agent[1]}")
        else:
            print("No tiene agente registrado.")
    else:
        # Ejecutar scraping para todos los jugadores
        scrape_players_from_database()
